src/crimson/os/seastore/nvmedevice/nvmedevice.h

```diff
@@ -0,0 +1,320 @@
+ * and SPDK, are available.
+ */
+class NVMeBlockDevice {
+protected:
+  uint64_t size = 0;
+  uint64_t block_size = 4096;
+
+  uint64_t write_granularity = 4096;
+  uint64_t write_alignment = 4096;
+
+public:
+  NVMeBlockDevice() {}
+  virtual ~NVMeBlockDevice() = default;
+
+  template <typename T>
+  static std::unique_ptr<T> create() {
+    return std::make_unique<T>();
+  }
+
+  /*
+   * Service NVMe device relative size
+   *
+   * size : total size of device in byte.
+   *
+   * block_size : IO unit size in byte. Caller should follow every IO command
+   * aligned with block size.
+   *
+   * preffered_write_granularity(PWG), preffered_write_alignment(PWA) : IO unit
+   * size for write in byte. Caller should request every write IO sized multiple
+   * times of PWG and aligned starting address by PWA. Available only if NVMe
+   * Device supports NVMe protocol 1.4 or later versions.
+   */
+  uint64_t get_size() const { return size; }
+  uint64_t get_block_size() const { return block_size; }
+
+  uint64_t get_preffered_write_granularity() const { return write_granularity; }
+  uint64_t get_preffered_write_alignment() const { return write_alignment; }
+
+  virtual read_ertr::future<> read(
+    uint64_t offset,
+    bufferptr &bptr) = 0;
+
+  /*
+   * Multi-stream write
+   *
+   * Give hint to device about classification of data whose life time is similar
+   * with each other. Data with same stream value will be managed together in
+   * SSD for better write performance.
+   */
+  virtual write_ertr::future<> write(
+    uint64_t offset,
+    bufferptr &bptr,
+    uint16_t stream = 0) = 0;
+
+  // TODO
+  virtual discard_ertr::future<> discard(
+    uint64_t offset,
+    uint64_t len) { return seastar::now(); }
+
+  virtual open_ertr::future<> open(
+      const std::string& path, 
+      seastar::open_flags mode) = 0;
+  virtual seastar::future<> close() = 0;
+
+  /*
+   * For passsing through nvme IO or Admin command to SSD
+   * Caller can construct and execute its own nvme command
+   */
+  virtual nvme_command_ertr::future<> pass_through_io(
+    NVMePassThroughCommand& command) { return nvme_command_ertr::now(); }
+  virtual nvme_command_ertr::future<> pass_admin(
+    nvme_admin_command_t& command) { return nvme_command_ertr::now(); }
+
+  /*
+   * End-to-End Data Protection
+   *
+   * NVMe device keeps track of data integrity similar with checksum. Client can
+   * offload checksuming to NVMe device to reduce its CPU utilization
+   */
+   virtual write_ertr::future<> protected_write(
+    uint64_t offset,
+    bufferptr &bptr,
+    uint16_t stream = 0) { return write_ertr::now(); }
+
+  /*
+   * Data Health
+   *
+   * Returns list of LBAs which have almost corrupted data. Data of the LBAs
+   * will be corrupted very soon. Caller can overwrite, unmap or refresh data to
+   * protect data
+   */
+   virtual nvme_command_ertr::future<> get_data_health(
+     std::list<uint64_t>& fragile_lbas) { return nvme_command_ertr::now(); }
```

Comment:
Why not return nvme_command_ertr::future<std::List<uint64_t>>?

Reply:
Fixed.